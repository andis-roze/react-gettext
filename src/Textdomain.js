import { Component } from 'react';
import PropTypes from 'prop-types';

class Textdomain extends Component {

	getChildContext() {
		const self = this;
		return {
			gettext: self.gettext,
			xgettext: self.xgettext,
			ngettext: self.ngettext,
			nxgettext: self.nxgettext,
		};
	}

	getTranslations() {
		const { catalog } = this.props;
		return typeof catalog === 'function' ? catalog() : catalog;
	}

	getPluralForm(n) {
		const { plural } = this.props;

		// return 0 if n is not integer
		if (isNaN(parseInt(n, 10))) {
			return 0;
		}

		// if pluralForm is function, use it to get plural form index
		if (typeof plural === 'function') {
			return plural(n);
		}

		// if pluralForm is string and contains only "n", "0-9", " ", "!=?:%+-/*><&|()"
		// characters, then we can "eval" it to calculate plural form
		if (typeof plural === 'string' && !plural.match(/[^n0-9 !=?:%+-/*><&|()]/i)) {
			/* eslint-disable no-new-func */
			const fnc = Function('n', `return ${plural}`);
			/* eslint-enable no-new-func */
			return +fnc(n);
		}

		return 0;
	}

	getDelimiter() {
		return '\u0004'; // End of Transmission (EOT)
	}

	gettext(message) {
		const messages = this.getTranslations();
		return Object.prototype.hasOwnProperty.call(messages, message)
			? messages[message]
			: message;
	}

	ngettext(singular, plural, n) {
		const self = this;
		const messages = self.getTranslations();
		const pluralIndex = self.getPluralForm(n);
		const defaultValue = n > 1 ? plural : singular;

		return Object.prototype.hasOwnProperty.call(messages, singular)
				&& Array.isArray(messages[singular])
				&& messages[singular].length >= pluralIndex
				&& pluralIndex >= 0
			? messages[singular][pluralIndex]
			: defaultValue;
	}

	xgettext(message, context) {
		const self = this;
		const EOT = self.getDelimiter();
		const messages = self.getTranslations();
		const key = context + EOT + message;

		return Object.prototype.hasOwnProperty.call(messages, key)
			? messages[key]
			: message;
	}

	nxgettext(singular, plural, n, context) {
		const self = this;
		const messages = self.getTranslations();
		const pluralIndex = self.getPluralForm(n);
		const defaultValue = n > 1 ? plural : singular;
		const EOT = self.getDelimiter();
		const key = context + EOT + singular;

		return Object.prototype.hasOwnProperty.call(messages, key)
				&& Array.isArray(messages[key])
				&& messages[key].length >= pluralIndex
				&& pluralIndex >= 0
			? messages[key][pluralIndex]
			: defaultValue;
	}

	render() {
		return this.props.children;
	}

}

Textdomain.propTypes = {
	catalog: PropTypes.oneOfType([
		PropTypes.func,
		PropTypes.objectOf(PropTypes.string),
	]),
	plural: PropTypes.oneOfType([
		PropTypes.func,
		PropTypes.string,
	]),
	children: PropTypes.arrayOf(PropTypes.node),
};

Textdomain.defaultProps = {
	catalog: {},
	plural: 'n != 1',
	children: [],
};

Textdomain.childContextTypes = {
	gettext: PropTypes.func,
	ngettext: PropTypes.func,
	xgettext: PropTypes.func,
	nxgettext: PropTypes.func,
};

export default Textdomain;
