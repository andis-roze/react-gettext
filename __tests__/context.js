import React, { Component } from 'react';
import PropTypes from 'prop-types';
import faker from 'faker';
import renderer from 'react-test-renderer';

import Textdomain from '../lib/index';

const SALT = '7[|UyK-%T`CKw-%j $1/6XTzdd(8+cY,/JCngNVv+wQO6NAv:+^=4GA[,+G@^GI<';

class App extends Component {
	render() {
		return <C1 {...this.props} />;
	}
};

class C1 extends Component {
	render() {
		return <span>{this.context.gettext(this.props.text)}</span>;
	}
};

C1.contextTypes = {
	gettext: PropTypes.func.isRequired,
	xgettext: PropTypes.func.isRequired,
	ngettext: PropTypes.func.isRequired
};

test('Child component has context functions', () => {
	const po = {};
	const sentence = faker.lorem.sentence();

	po[sentence] = SALT;

	const hoc = Textdomain(po, 'n!=1')(App);
	const component = renderer.create(React.createElement(hoc, {text: sentence}, []));

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});