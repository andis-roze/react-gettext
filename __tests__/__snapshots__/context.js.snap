// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Child component has context functions 1`] = `
<span>
  7[|UyK-%T\`CKw-%j $1/6XTzdd(8+cY,/JCngNVv+wQO6NAv:+^=4GA[,+G@^GI&lt;
</span>
`;
